File: packages/sdk-ts/src/decisions/verify.ts
/**
 * Offline verification of decision grants (PRD G-3).
 *
 * Verification proves that a decision grant is authentic, unexpired and
 * approves exactly the action about to be performed on the current case
 * version, and, for four eyes, that two different people approved it. It does
 * not prove that the grant has not been used: that needs the issuer's atomic
 * consumption (`grantex.decisions.consume()`), which `enforce()` performs
 * before it allows the call. The Python SDK implements the same rules
 * (`grantex.decisions`); the token profile is `spec/decision-grant.md`.
 */
import { createRemoteJWKSet, decodeProtectedHeader, jwtVerify, type JWTPayload, type JWTVerifyGetKey } from 'jose';
import { DecisionSubReason } from '../denials.js';
import { ActionValidationError, computeActionHash, isActionHash, parseDecisionAction, type DecisionAction } from './action.js';

export const DECISION_GRANT_TYP = 'decision+jwt';
export const DECISION_GRANT_AUDIENCE = 'urn:grantex:decision';

const JTI_RE = /^dgnt_[0-9A-HJKMNP-TV-Z]{26}$/;
const MAX_TOKEN_LENGTH = 16_384;

/** A decision grant is absent, unusable or does not match the call. */
export class DecisionGrantError extends Error {
  readonly subReason: DecisionSubReason;
  constructor(subReason: DecisionSubReason, message: string) {
    super(message);
    this.name = 'DecisionGrantError';
    this.subReason = subReason;
  }
}

export interface FourEyes {
  approvalsRequired: 2;
  position: 1 | 2;
  firstJti?: string;
  firstSub?: string;
}

/** The verified claims of one decision grant. */
export interface DecisionGrant {
  token: string;
  jti: string;
  iss: string;
  sub: string;
  dev: string;
  idp: string;
  approverAuth: string;
  acr?: string;
  amr: string[];
  authTime: number;
  action: DecisionAction;
  actionHash: string;
  connector: string;
  caseVersion: string;
  dwellMs: number;
  /** Always `server`: the issuer measured the dwell time. */
  dwellSource: 'server';
  decisionRequest: string;
  memoHash: string;
  policyScoreHash: string;
  iat: number;
  exp: number;
  memoRef?: string;
  policyScoreRef?: string;
  fourEyes?: FourEyes;
}

/** Decision grants that together authorise one action. */
export interface DecisionGrantSet {
  grants: DecisionGrant[];
  action: DecisionAction;
  actionHash: string;
  caseVersion: string;
  approvalsRequired: 1 | 2;
}

/** A verification key, or a resolver that picks one from the protected header. */
export type DecisionKey = CryptoKey | Uint8Array | JWTVerifyGetKey;

export interface VerifyDecisionGrantOptions {
  /** Expected `iss`. */
  issuer: string;
  /** JWKS URL of the issuer; used when `key` is not given. */
  jwksUri?: string;
  /** Verification key or resolver; overrides `jwksUri`. */
  key?: DecisionKey;
  /** The grant must have been issued to this developer (`dev`). */
  developerId?: string;
  /** The grant must be for this connector. */
  connector?: string;
  /** Accepted algorithms, a subset of RS256 and ES256. Default both. */
  algorithms?: string[];
  /** Seconds of clock tolerance. Default 0. */
  clockTolerance?: number;
  /** Current time in seconds (for tests). */
  now?: number;
}

export interface VerifyDecisionGrantsOptions extends VerifyDecisionGrantOptions {
  /** 2 when the manifest lists the decision in `four_eyes_on`. Default 1. */
  approvalsRequired?: 1 | 2;
}

const MAX_JWKS_RESOLVERS = 64;
const jwksResolvers = new Map<string, JWTVerifyGetKey>();

function remoteKey(jwksUri: string): JWTVerifyGetKey {
  const url = new URL(jwksUri);
  url.hash = '';
  const cached = jwksResolvers.get(url.href);
  if (cached) return cached;
  const resolver = createRemoteJWKSet(url);
  if (jwksResolvers.size >= MAX_JWKS_RESOLVERS) {
    const oldest = jwksResolvers.keys().next().value as string | undefined;
    if (oldest !== undefined) jwksResolvers.delete(oldest);
  }
  jwksResolvers.set(url.href, resolver);
  return resolver;
}

function toAction(value: unknown): DecisionAction {
  try {
    return parseDecisionAction(value);
  } catch (err) {
    if (err instanceof ActionValidationError) {
      throw new DecisionGrantError(DecisionSubReason.MALFORMED, `expected action is invalid: ${err.message}`);
    }
    throw err;
  }
}

function claimString(payload: JWTPayload, key: string): string {
  const value = (payload as Record<string, unknown>)[key];
  if (typeof value !== 'string' || value.length === 0) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, `decision grant claim ${key} is missing`);
  }
  return value;
}

function claimInt(payload: JWTPayload, key: string): number {
  const value = (payload as Record<string, unknown>)[key];
  if (typeof value !== 'number' || !Number.isSafeInteger(value) || value < 0) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, `decision grant claim ${key} is invalid`);
  }
  return value;
}

function parseFourEyes(value: unknown): FourEyes | undefined {
  if (value === undefined) return undefined;
  const v = value as Record<string, unknown>;
  if (typeof value !== 'object' || value === null || Array.isArray(value) || v['approvals_required'] !== 2 || (v['position'] !== 1 && v['position'] !== 2)) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant four_eyes is invalid');
  }
  const firstJti = v['first_jti'];
  const firstSub = v['first_sub'];
  if (v['position'] === 2 && (typeof firstJti !== 'string' || typeof firstSub !== 'string')) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'the second approval must name the first');
  }
  return {
    approvalsRequired: 2,
    position: v['position'],
    ...(typeof firstJti === 'string' ? { firstJti } : {}),
    ...(typeof firstSub === 'string' ? { firstSub } : {}),
  };
}

/**
 * Verifies one decision grant against the action about to be performed.
 * Refusals (`DecisionGrantError.subReason`): `malformed` (type, signature,
 * issuer, audience, claims, or an `action_hash` that is not the hash of the
 * `action` claim), `unknown_grant` (another developer), `wrong_case`,
 * `action_mismatch` (another action or connector), `case_changed`, `expired`.
 * Single use is not checked here.
 */
export async function verifyDecisionGrant(
  token: string,
  expectedAction: DecisionAction,
  caseVersion: string,
  options: VerifyDecisionGrantOptions,
): Promise<DecisionGrant> {
  if (typeof token !== 'string' || token.length === 0 || token.length > MAX_TOKEN_LENGTH) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant must be a compact JWT');
  }
  const expected = toAction(expectedAction);
  if (typeof caseVersion !== 'string' || caseVersion.length === 0) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'caseVersion is required');
  }
  const algorithms: string[] = (options.algorithms ?? ['RS256', 'ES256']).filter((a) => a === 'RS256' || a === 'ES256');
  let header;
  try {
    header = decodeProtectedHeader(token);
  } catch {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant is not a JWT');
  }
  if (header.typ !== DECISION_GRANT_TYP) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, `decision grant typ must be ${DECISION_GRANT_TYP}`);
  }
  if (typeof header.kid !== 'string' || header.kid.length === 0) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant has no kid');
  }
  if (typeof header.alg !== 'string' || !algorithms.includes(header.alg)) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, `decision grant algorithm ${String(header.alg)} is not allowed`);
  }
  const key = options.key ?? (options.jwksUri !== undefined ? remoteKey(options.jwksUri) : undefined);
  if (key === undefined) throw new Error('verifyDecisionGrant needs jwksUri or key');

  let payload: JWTPayload;
  try {
    const verifyOptions = {
      issuer: options.issuer,
      audience: DECISION_GRANT_AUDIENCE,
      algorithms,
      typ: DECISION_GRANT_TYP,
      requiredClaims: ['iss', 'aud', 'sub', 'jti', 'iat', 'exp'],
      // Time is checked below against `now`, after the signature.
      currentDate: new Date(0),
    };
    ({ payload } = typeof key === 'function'
      ? await jwtVerify(token, key, verifyOptions)
      : await jwtVerify(token, key, verifyOptions));
  } catch (err) {
    throw new DecisionGrantError(
      DecisionSubReason.MALFORMED,
      `decision grant signature, issuer or audience is invalid: ${err instanceof Error ? err.message : 'verification failed'}`,
    );
  }

  const jti = claimString(payload, 'jti');
  if (!JTI_RE.test(jti)) throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant jti is malformed');
  let action: DecisionAction;
  try {
    action = parseDecisionAction(payload['action']);
  } catch {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant action is malformed');
  }
  const actionHash = claimString(payload, 'action_hash');
  if (!isActionHash(actionHash) || computeActionHash(action) !== actionHash) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant action_hash does not match its action');
  }
  const amr = payload['amr'];
  if (!Array.isArray(amr) || !amr.every((v) => typeof v === 'string')) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant amr is invalid');
  }
  const acr = payload['acr'];
  if (acr !== undefined && typeof acr !== 'string') {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant acr is invalid');
  }
  if (payload['dwell_source'] !== 'server') {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant dwell time was not measured by the issuer');
  }
  const memoHash = claimString(payload, 'memo_hash');
  const policyScoreHash = claimString(payload, 'policy_score_hash');
  if (!isActionHash(memoHash) || !isActionHash(policyScoreHash)) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant memo or policy score hash is malformed');
  }
  const iat = claimInt(payload, 'iat');
  const exp = claimInt(payload, 'exp');
  if (exp <= iat || exp - iat > 86_400) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant lifetime exceeds 24 hours');
  }
  const memoRef = payload['memo_ref'];
  const policyScoreRef = payload['policy_score_ref'];
  const fourEyes = parseFourEyes(payload['four_eyes']);
  const grant: DecisionGrant = {
    token,
    jti,
    iss: claimString(payload, 'iss'),
    sub: claimString(payload, 'sub'),
    dev: claimString(payload, 'dev'),
    idp: claimString(payload, 'idp'),
    approverAuth: claimString(payload, 'approver_auth'),
    ...(typeof acr === 'string' ? { acr } : {}),
    amr: amr as string[],
    authTime: claimInt(payload, 'auth_time'),
    action,
    actionHash,
    connector: claimString(payload, 'connector'),
    caseVersion: claimString(payload, 'case_version'),
    dwellMs: claimInt(payload, 'dwell_ms'),
    dwellSource: 'server',
    decisionRequest: claimString(payload, 'decision_request'),
    memoHash,
    policyScoreHash,
    iat,
    exp,
    ...(typeof memoRef === 'string' ? { memoRef } : {}),
    ...(typeof policyScoreRef === 'string' ? { policyScoreRef } : {}),
    ...(fourEyes !== undefined ? { fourEyes } : {}),
  };

  if (options.developerId !== undefined && grant.dev !== options.developerId) {
    throw new DecisionGrantError(DecisionSubReason.UNKNOWN_GRANT, 'decision grant was issued to another developer');
  }
  if (grant.action.case_id !== expected.case_id) {
    throw new DecisionGrantError(DecisionSubReason.WRONG_CASE, 'decision grant is for another case');
  }
  if (grant.actionHash !== computeActionHash(expected)) {
    throw new DecisionGrantError(DecisionSubReason.ACTION_MISMATCH, 'decision grant approves a different action');
  }
  if (options.connector !== undefined && grant.connector !== options.connector) {
    throw new DecisionGrantError(DecisionSubReason.ACTION_MISMATCH, 'decision grant approves an action on another connector');
  }
  if (grant.caseVersion !== caseVersion) {
    throw new DecisionGrantError(DecisionSubReason.CASE_CHANGED, 'decision grant was approved for another case version');
  }
  const now = options.now ?? Math.floor(Date.now() / 1000);
  const tolerance = options.clockTolerance ?? 0;
  if (grant.iat > now + tolerance + 60) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grant is issued in the future');
  }
  if (now >= grant.exp + tolerance) {
    throw new DecisionGrantError(DecisionSubReason.EXPIRED, 'decision grant has expired');
  }
  return grant;
}

/**
 * Verifies the decision grants for one action, including four eyes: with two
 * approvals required (from the manifest, or because a grant says so) exactly
 * two grants with different `sub` are needed, the second naming the first.
 */
export async function verifyDecisionGrants(
  tokens: readonly string[],
  expectedAction: DecisionAction,
  caseVersion: string,
  options: VerifyDecisionGrantsOptions,
): Promise<DecisionGrantSet> {
  const approvalsRequired = options.approvalsRequired ?? 1;
  if (approvalsRequired !== 1 && approvalsRequired !== 2) throw new Error('approvalsRequired must be 1 or 2');
  if (!Array.isArray(tokens)) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'decision grants must be an array of tokens');
  }
  if (tokens.length === 0) throw new DecisionGrantError(DecisionSubReason.ABSENT, 'no decision grant was presented');
  if (tokens.length > 2) throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'at most two decision grants can be presented');
  const expected = toAction(expectedAction);
  const grants: DecisionGrant[] = [];
  for (const token of tokens) grants.push(await verifyDecisionGrant(token, expected, caseVersion, options));

  const required: 1 | 2 = grants.some((g) => g.fourEyes !== undefined) ? 2 : approvalsRequired;
  if (new Set(grants.map((g) => g.jti)).size !== grants.length) {
    throw new DecisionGrantError(DecisionSubReason.SAME_APPROVER, 'the same decision grant was presented twice');
  }
  if (grants.length < required) {
    throw new DecisionGrantError(DecisionSubReason.FOUR_EYES_INCOMPLETE, 'this decision needs two approvals from different people');
  }
  if (grants.length > required) {
    throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'more decision grants than this decision needs');
  }
  let ordered = grants;
  if (required === 2) {
    const [first, second] = [...grants].sort((a, b) => (a.fourEyes?.position ?? 0) - (b.fourEyes?.position ?? 0)) as [DecisionGrant, DecisionGrant];
    if (first.sub === second.sub) {
      throw new DecisionGrantError(DecisionSubReason.SAME_APPROVER, 'both decision grants were approved by the same person');
    }
    if (first.fourEyes?.position !== 1 || second.fourEyes?.position !== 2
        || second.fourEyes.firstJti !== first.jti || second.fourEyes.firstSub !== first.sub
        || first.decisionRequest !== second.decisionRequest) {
      throw new DecisionGrantError(DecisionSubReason.MALFORMED, 'the second approval does not reference the first');
    }
    ordered = [first, second];
  }
  return { grants: ordered, action: expected, actionHash: computeActionHash(expected), caseVersion, approvalsRequired: required };
}

File: packages/sdk-py/src/grantex/decisions/_verify.py
"""Offline verification of decision grants (PRD G-3).

Verification here proves that a decision grant is authentic, unexpired and
approves exactly the action about to be performed on the current case
version, and, for four eyes, that two different people approved it. It does
**not** prove that the grant has not been used: that needs the issuer's
atomic consumption (:meth:`grantex.Grantex.decisions.consume`), which
``enforce()`` performs before it allows the call.

The token profile is specified in ``spec/decision-grant.md``.
"""

from __future__ import annotations

import re
import time
from dataclasses import dataclass
from typing import Any, Callable, Mapping, Optional, Sequence, Tuple, Union

import jwt

from .._errors import GrantexTokenError
from ..denials import DecisionSubReason
from ._action import ActionValidationError, DecisionAction, is_action_hash

__all__ = [
    "DECISION_GRANT_AUDIENCE",
    "DECISION_GRANT_TYP",
    "DecisionGrant",
    "DecisionGrantError",
    "DecisionGrantSet",
    "FourEyes",
    "verify_decision_grant",
    "verify_decision_grants",
]

DECISION_GRANT_TYP = "decision+jwt"
DECISION_GRANT_AUDIENCE = "urn:grantex:decision"

_JTI_RE = re.compile(r"^dgnt_[0-9A-HJKMNP-TV-Z]{26}\Z")
_MAX_TOKEN_LENGTH = 16_384

KeyResolver = Callable[[Mapping[str, Any]], Any]
"""Returns the verification key for a token's protected header."""


class DecisionGrantError(Exception):
    """A decision grant is absent, unusable or does not match the call.

    ``sub_reason`` is a :class:`grantex.denials.DecisionSubReason` value.
    """

    def __init__(self, sub_reason: str, message: str) -> None:
        super().__init__(message)
        self.sub_reason = sub_reason


@dataclass(frozen=True)
class FourEyes:
    approvals_required: int
    position: int
    first_jti: Optional[str] = None
    first_sub: Optional[str] = None


@dataclass(frozen=True)
class DecisionGrant:
    """The verified claims of one decision grant."""

    token: str
    jti: str
    iss: str
    sub: str
    dev: str
    idp: str
    approver_auth: str
    amr: Tuple[str, ...]
    auth_time: int
    action: DecisionAction
    action_hash: str
    connector: str
    case_version: str
    dwell_ms: int
    decision_request: str
    iat: int
    exp: int
    memo_hash: str
    policy_score_hash: str
    dwell_source: str = "server"
    acr: Optional[str] = None
    memo_ref: Optional[str] = None
    policy_score_ref: Optional[str] = None
    four_eyes: Optional[FourEyes] = None


@dataclass(frozen=True)
class DecisionGrantSet:
    """Decision grants that together authorise one action."""

    grants: Tuple[DecisionGrant, ...]
    action: DecisionAction
    action_hash: str
    case_version: str
    approvals_required: int

    @property
    def tokens(self) -> Tuple[str, ...]:
        return tuple(g.token for g in self.grants)

    @property
    def jtis(self) -> Tuple[str, ...]:
        return tuple(g.jti for g in self.grants)


def _as_action(value: Union[DecisionAction, Mapping[str, Any]]) -> DecisionAction:
    if isinstance(value, DecisionAction):
        return value
    try:
        return DecisionAction.from_dict(value)
    except ActionValidationError as exc:
        raise DecisionGrantError(
            DecisionSubReason.MALFORMED, f"expected action is invalid: {exc}"
        ) from exc


_KEY_TYPES = {"RS256": "RSA", "ES256": "EC"}
_UNKNOWN_KID_COOLDOWN_SECONDS = 30.0


def _default_key_resolver(jwks_uri: str) -> KeyResolver:
    """Resolves the key named by ``kid`` in the issuer's JWKS, of the type the
    algorithm needs (RSA for RS256, P-256 EC for ES256); an unknown ``kid``
    refetches the set at most once per cooldown."""
    from jwt.algorithms import ECAlgorithm, RSAAlgorithm

    from .._verify import _get_jwks

    def resolve(header: Mapping[str, Any]) -> Any:
        kid = header.get("kid")
        alg = header.get("alg")
        if not isinstance(kid, str) or not kid:
            raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant has no kid")
        kty = _KEY_TYPES.get(alg) if isinstance(alg, str) else None
        if kty is None:
            raise DecisionGrantError(DecisionSubReason.MALFORMED, f"decision grant algorithm {alg!r} is not allowed")

        def select(keys: Sequence[Mapping[str, Any]]) -> Optional[Mapping[str, Any]]:
            matches = [k for k in keys if k.get("kid") == kid and k.get("kty") == kty]
            if len(matches) > 1:
                raise DecisionGrantError(DecisionSubReason.MALFORMED, f"JWKS has several keys with kid {kid!r}")
            if matches and kty == "EC" and matches[0].get("crv") != "P-256":
                return None
            return matches[0] if matches else None

        try:
            entry = _get_jwks(jwks_uri)
            matched = select(entry.keys)
            if matched is None and time.monotonic() - entry.fetched_at >= _UNKNOWN_KID_COOLDOWN_SECONDS:
                entry = _get_jwks(jwks_uri, force_refresh=True)
                matched = select(entry.keys)
        except GrantexTokenError as exc:
            raise DecisionGrantError(DecisionSubReason.MALFORMED, f"issuer keys unavailable: {exc}") from exc
        if matched is None:
            raise DecisionGrantError(DecisionSubReason.MALFORMED, f"no {kty} key with kid {kid!r} in the issuer's JWKS")
        try:
            return (RSAAlgorithm if kty == "RSA" else ECAlgorithm).from_jwk(dict(matched))
        except Exception as exc:  # noqa: BLE001 - an unusable key refuses the grant
            raise DecisionGrantError(DecisionSubReason.MALFORMED, "issuer key cannot be used") from exc

    return resolve


def _claim_str(payload: Mapping[str, Any], key: str) -> str:
    value = payload.get(key)
    if not isinstance(value, str) or not value:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, f"decision grant claim {key} is missing")
    return value


def _claim_int(payload: Mapping[str, Any], key: str) -> int:
    value = payload.get(key)
    if isinstance(value, bool) or not isinstance(value, int) or value < 0:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, f"decision grant claim {key} is invalid")
    return value


def _parse_four_eyes(value: Any) -> Optional[FourEyes]:
    if value is None:
        return None
    if not isinstance(value, Mapping) or value.get("approvals_required") != 2 or value.get("position") not in (1, 2):
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant four_eyes is invalid")
    first_jti = value.get("first_jti")
    first_sub = value.get("first_sub")
    if value["position"] == 2 and (not isinstance(first_jti, str) or not isinstance(first_sub, str)):
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "the second approval must name the first")
    return FourEyes(
        approvals_required=2,
        position=int(value["position"]),
        first_jti=first_jti if isinstance(first_jti, str) else None,
        first_sub=first_sub if isinstance(first_sub, str) else None,
    )


def verify_decision_grant(
    token: str,
    expected_action: Union[DecisionAction, Mapping[str, Any]],
    case_version: str,
    *,
    jwks_uri: Optional[str] = None,
    issuer: str,
    key_resolver: Optional[KeyResolver] = None,
    developer_id: Optional[str] = None,
    connector: Optional[str] = None,
    algorithms: Sequence[str] = ("RS256", "ES256"),
    clock_tolerance: int = 0,
    now: Optional[int] = None,
) -> DecisionGrant:
    """Verify one decision grant against the action about to be performed.

    Checks, failing with :class:`DecisionGrantError` and a sub-reason:

    - ``malformed``: not a ``typ: decision+jwt`` JWT, bad signature, wrong
      issuer or audience, missing or invalid claims, or an ``action_hash`` that
      is not the hash of the ``action`` claim;
    - ``unknown_grant``: issued to another developer (``developer_id``);
    - ``wrong_case``: approves an action on another case;
    - ``action_mismatch``: approves another action on this case, or for
      another connector;
    - ``case_changed``: approved for another case version;
    - ``expired``: past ``exp`` (after the signature is known to be good).

    It does not check single use; see the module documentation.
    """
    if not isinstance(token, str) or not token or len(token) > _MAX_TOKEN_LENGTH:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant must be a compact JWT")
    expected = _as_action(expected_action)
    if not isinstance(case_version, str) or not case_version:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "case_version is required")
    if key_resolver is None:
        if jwks_uri is None:
            raise ValueError("verify_decision_grant needs jwks_uri or key_resolver")
        key_resolver = _default_key_resolver(jwks_uri)

    try:
        header = jwt.get_unverified_header(token)
    except jwt.PyJWTError as exc:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, f"decision grant is not a JWT: {exc}") from exc
    if header.get("typ") != DECISION_GRANT_TYP:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, f"decision grant typ must be {DECISION_GRANT_TYP}")
    if not isinstance(header.get("kid"), str) or not header.get("kid"):
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant has no kid")
    if header.get("alg") not in tuple(algorithms) or header.get("alg") not in _KEY_TYPES:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, f"decision grant algorithm {header.get('alg')!r} is not allowed")
    try:
        key = key_resolver(header)
    except DecisionGrantError:
        raise
    except GrantexTokenError as exc:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, f"no key to verify the decision grant: {exc}") from exc

    try:
        payload: dict[str, Any] = jwt.decode(
            token,
            key,
            algorithms=list(algorithms),
            audience=DECISION_GRANT_AUDIENCE,
            issuer=issuer,
            options={
                # Time is checked below against ``now``, after the signature.
                "verify_exp": False, "verify_iat": False, "verify_nbf": False,
                "require": ["iss", "aud", "sub", "jti", "iat", "exp"],
            },
        )
    except jwt.PyJWTError as exc:
        raise DecisionGrantError(
            DecisionSubReason.MALFORMED, f"decision grant signature, issuer or audience is invalid: {exc}"
        ) from exc

    jti = _claim_str(payload, "jti")
    if not _JTI_RE.match(jti):
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant jti is malformed")
    try:
        action = DecisionAction.from_dict(payload.get("action"))
    except ActionValidationError as exc:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, f"decision grant action is malformed: {exc}") from exc
    action_hash = _claim_str(payload, "action_hash")
    if not is_action_hash(action_hash) or action.action_hash() != action_hash:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant action_hash does not match its action")
    amr = payload.get("amr")
    if not isinstance(amr, list) or not all(isinstance(v, str) for v in amr):
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant amr is invalid")
    acr = payload.get("acr")
    if acr is not None and not isinstance(acr, str):
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant acr is invalid")
    if payload.get("dwell_source") != "server":
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant dwell time was not measured by the issuer")
    memo_hash = _claim_str(payload, "memo_hash")
    policy_score_hash = _claim_str(payload, "policy_score_hash")
    if not is_action_hash(memo_hash) or not is_action_hash(policy_score_hash):
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant memo or policy score hash is malformed")
    iat = _claim_int(payload, "iat")
    exp = _claim_int(payload, "exp")
    if exp <= iat or exp - iat > 86_400:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant lifetime exceeds 24 hours")

    grant = DecisionGrant(
        token=token,
        jti=jti,
        iss=_claim_str(payload, "iss"),
        sub=_claim_str(payload, "sub"),
        dev=_claim_str(payload, "dev"),
        idp=_claim_str(payload, "idp"),
        approver_auth=_claim_str(payload, "approver_auth"),
        amr=tuple(amr),
        auth_time=_claim_int(payload, "auth_time"),
        action=action,
        action_hash=action_hash,
        connector=_claim_str(payload, "connector"),
        case_version=_claim_str(payload, "case_version"),
        dwell_ms=_claim_int(payload, "dwell_ms"),
        decision_request=_claim_str(payload, "decision_request"),
        iat=iat,
        exp=exp,
        memo_hash=memo_hash,
        policy_score_hash=policy_score_hash,
        acr=acr,
        memo_ref=payload.get("memo_ref") if isinstance(payload.get("memo_ref"), str) else None,
        policy_score_ref=payload.get("policy_score_ref") if isinstance(payload.get("policy_score_ref"), str) else None,
        four_eyes=_parse_four_eyes(payload.get("four_eyes")),
    )

    if developer_id is not None and grant.dev != developer_id:
        raise DecisionGrantError(DecisionSubReason.UNKNOWN_GRANT, "decision grant was issued to another developer")
    if grant.action.case_id != expected.case_id:
        raise DecisionGrantError(DecisionSubReason.WRONG_CASE, "decision grant is for another case")
    if grant.action_hash != expected.action_hash():
        raise DecisionGrantError(DecisionSubReason.ACTION_MISMATCH, "decision grant approves a different action")
    if connector is not None and grant.connector != connector:
        raise DecisionGrantError(DecisionSubReason.ACTION_MISMATCH, "decision grant approves an action on another connector")
    if grant.case_version != case_version:
        raise DecisionGrantError(DecisionSubReason.CASE_CHANGED, "decision grant was approved for another case version")
    current = int(time.time()) if now is None else now
    if grant.iat > current + clock_tolerance + 60:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grant is issued in the future")
    if current >= grant.exp + clock_tolerance:
        raise DecisionGrantError(DecisionSubReason.EXPIRED, "decision grant has expired")
    return grant


def verify_decision_grants(
    tokens: Sequence[str],
    expected_action: Union[DecisionAction, Mapping[str, Any]],
    case_version: str,
    *,
    issuer: str,
    approvals_required: int = 1,
    jwks_uri: Optional[str] = None,
    key_resolver: Optional[KeyResolver] = None,
    developer_id: Optional[str] = None,
    connector: Optional[str] = None,
    algorithms: Sequence[str] = ("RS256", "ES256"),
    clock_tolerance: int = 0,
    now: Optional[int] = None,
) -> DecisionGrantSet:
    """Verify the decision grants for one action, including four eyes.

    ``approvals_required`` is 2 when the manifest lists the decision in
    ``four_eyes_on``; a grant that itself says two approvals are required
    raises the requirement to 2 as well. With two approvals required, exactly
    two grants with different ``sub`` are needed, the second naming the first
    (``same_approver``, ``four_eyes_incomplete`` or ``malformed`` otherwise).
    """
    if approvals_required not in (1, 2):
        raise ValueError("approvals_required must be 1 or 2")
    if isinstance(tokens, str) or not isinstance(tokens, Sequence):
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "decision grants must be a sequence of tokens")
    if len(tokens) == 0:
        raise DecisionGrantError(DecisionSubReason.ABSENT, "no decision grant was presented")
    if len(tokens) > 2:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "at most two decision grants can be presented")
    expected = _as_action(expected_action)
    grants = tuple(
        verify_decision_grant(
            t, expected, case_version,
            jwks_uri=jwks_uri, issuer=issuer, key_resolver=key_resolver, developer_id=developer_id,
            connector=connector, algorithms=algorithms, clock_tolerance=clock_tolerance, now=now,
        )
        for t in tokens
    )
    required = max([approvals_required] + [g.four_eyes.approvals_required for g in grants if g.four_eyes])
    if len({g.jti for g in grants}) != len(grants):
        raise DecisionGrantError(DecisionSubReason.SAME_APPROVER, "the same decision grant was presented twice")
    if len(grants) < required:
        raise DecisionGrantError(
            DecisionSubReason.FOUR_EYES_INCOMPLETE,
            "this decision needs two approvals from different people",
        )
    if len(grants) > required:
        raise DecisionGrantError(DecisionSubReason.MALFORMED, "more decision grants than this decision needs")
    if required == 2:
        first, second = sorted(grants, key=lambda g: g.four_eyes.position if g.four_eyes else 0)
        if first.sub == second.sub:
            raise DecisionGrantError(DecisionSubReason.SAME_APPROVER, "both decision grants were approved by the same person")
        if (
            first.four_eyes is None or second.four_eyes is None
            or first.four_eyes.position != 1 or second.four_eyes.position != 2
            or second.four_eyes.first_jti != first.jti or second.four_eyes.first_sub != first.sub
            or first.decision_request != second.decision_request
        ):
            raise DecisionGrantError(DecisionSubReason.MALFORMED, "the second approval does not reference the first")
        grants = (first, second)
    return DecisionGrantSet(
        grants=grants,
        action=expected,
        action_hash=expected.action_hash(),
        case_version=case_version,
        approvals_required=required,
    )
